Add impl `http_body::Body` for `hyper::Body`

LucioFranco opened the pull request:
This adds a `http_body::Body` impl for hypers `Body`. This should
allow us to start moving to a more generic body trait based on
`BufStream` and `http-body`.

cc @seanmonstar

LucioFranco commented:
GitCop 😭 idk what is wrong.

seanmonstar commented:
Don't worry about GitCop, I always fix it as part of the squash step.

LucioFranco commented:
@seanmonstar so do you suggest a blanket impl of Body for T: Payload? Otherwise those changes seem fine, will get to them soon.

seanmonstar commented:
No, I mean instead of having `poll_trailers` have duplicate code, do like this:

```rust
impl http_body::Body for Body {
    fn poll_trailers(&mut self) -> Poll<Option<HeaderMap>, Self::Error> {
        Payload::poll_trailers(self)
    }
    // etc...
}
```

LucioFranco commented:
@seanmonstar I think I've addressed all your comments. Should be good for another pass.

seanmonstar commented on src/body/body.rs:
```diff
@@ -332,6 +322,35 @@ impl Payload for Body {
+        let upper_limit = <Self as Payload>::content_length(self);
+
+        if let Some(upper) = upper_limit {
+            hint.set_upper(upper);
```
I believe we can set the lower as well, since it's meant to be an exact size in `Payload`.

LucioFranco replied:
so if I understand correctly, since we ahve the total content_length here so the exact size, lower would == upper?

seanmonstar replied:
Yep, `Payload::content_length` is an exact size if `Some(_)`.

seanmonstar commented on src/body/body.rs:
```diff
@@ -183,54 +183,44 @@ impl Body {
 
     fn poll_eof(&mut self) -> Poll<Option<Chunk>, ::Error> {
         match self.take_delayed_eof() {
-            Some(DelayEof::NotEof(mut delay)) => {
```
Just checking, this is simply rustfmt?

LucioFranco replied:
Ah damn totally could have been forgot to disable it for this PR oops. Would you like me to revert this?

LucioFranco commented:
@seanmonstar I addressed your comment and reverted the rustfmt. Only issue is that h2 introduced breaking changes for 1.26. We can try to pin it to an older version?

seanmonstar commented:
Just finished publishing a new h2 version that should work on Rust 1.26, so I gave CI a nudge.

seanmonstar commented:
Woot, green! Thanks!
